File: AudioUnitV3Example/Filter/OSX/FilterDemoFramework/FilterDemoViewController.m
/*
See LICENSE.txt for this sample’s licensing information.

Abstract:
View controller for the FilterDemo audio unit. Manages the interactions between a FilterView and the audio unit's parameters.
*/

#import <Cocoa/Cocoa.h>
#import <FilterDemoFramework/FilterDemo.h>
#import "FilterDemoViewController+AUAudioUnitFactory.h"
#import "FilterDemoViewController_Internal.h"
#import "FilterView.h"

@interface FilterDemoViewController () {

    // These are moved in and out of the container view, so we want to hold on to them
    __strong IBOutlet NSView* smallView;
    __strong IBOutlet NSView* largeView;

    __weak IBOutlet FilterView  *filterView;
    __weak IBOutlet NSTextField *frequencyLabel;
    __weak IBOutlet NSTextField *resonanceLabel;

    __weak IBOutlet NSTextField *frequencyLabelSmallView;
    __weak IBOutlet NSTextField *resonanceLabelSmallView;

    bool smallViewActive;
    AUParameter *cutoffParameter;
    AUParameter *resonanceParameter;
    AUParameterObserverToken parameterObserverToken;
}

-(IBAction) setCutoff:(id)sender;
-(IBAction) setResonance:(id)sender;
@end

@implementation FilterDemoViewController

#pragma mark <AUAudioUnitFactory>

/* 
 The principal class of a UI v3 audio unit extension must derive from AUViewController and implement the AUAudioUnitFactory protocol.
 
 - createAudioUnitWithComponentDescription:error: should create and return an instance of its audio unit.
 
 This method will be called only once per instance of the factory.
 
 Note that in non-ARC code, "create" methods return unretained objects (unlike "create" C functions); the implementor should return an object with reference count 1 but autoreleased.
*/

- (AUv3FilterDemo *)createAudioUnitWithComponentDescription:(AudioComponentDescription)desc
                                                      error:(NSError **)error {
    self.audioUnit = [[AUv3FilterDemo alloc] initWithComponentDescription:desc error:error];
    return self.audioUnit;
}

#pragma mark-
#pragma mark: AUViewController

- (id)init {
    self = [super initWithNibName:@"FilterDemoViewController"
                           bundle:[NSBundle bundleForClass:NSClassFromString(@"FilterDemoViewController")]];
    return self;
}

- (void)dealloc {
    filterView.delegate = nil;
    [self disconnectViewWithAU];
    NSLog(@"FilterDemoViewController Dealloc\n");
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    filterView.delegate = self;

    if (_audioUnit) {
        [self connectViewWithAU];
        _audioUnit.filterDemoViewController = self;
    }

    // We will set our constraints manually
    smallView.translatesAutoresizingMaskIntoConstraints = NO;
    self.view.translatesAutoresizingMaskIntoConstraints = NO;
    largeView.translatesAutoresizingMaskIntoConstraints = NO;

    smallViewActive = false;
}

#pragma mark-

- (AUv3FilterDemo *)getAudioUnit {
    return _audioUnit;
}

- (void)setAudioUnit:(AUv3FilterDemo *)audioUnit {
    _audioUnit = audioUnit;
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([self isViewLoaded]) {
            [self connectViewWithAU];
            _audioUnit.filterDemoViewController = self;
        }
    });
}

#pragma mark-

/*
 The "allParameterValues" pseudo-property is used to issue KVO notifications in response to certain events where potentially all parameter values are invalidated. This includes changes to currentPreset, fullState, and fullStateForDocument.
*/
- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary<NSString *, id> *)change
                       context:(void *)context {
    
    NSLog(@"FilterDemoViewControler allParameterValues key path changed: %s\n", keyPath.UTF8String);

    dispatch_async(dispatch_get_main_queue(), ^{
        
        filterView.frequency = cutoffParameter.value;
        filterView.resonance = resonanceParameter.value;
        
        [self updateLabelsWithFrequency:[cutoffParameter stringFromValue: nil] resonance:[resonanceParameter stringFromValue: nil]];
        [self updateFilterViewFrequencyAndMagnitudes];
    });
}

#pragma mark-

- (void)updateLabelsWithFrequency:(NSString *)frequencyString resonance:(NSString *)resonanceString {
    if (frequencyString) {
        frequencyLabel.stringValue = frequencyString;
        frequencyLabelSmallView.stringValue = frequencyString;
    }

    if (resonanceString) {
        resonanceLabel.stringValue = resonanceString;
        resonanceLabelSmallView.stringValue = resonanceString;
    }
}

- (void)connectViewWithAU {
    AUParameterTree *paramTree = _audioUnit.parameterTree;
    
    if (paramTree) {
        cutoffParameter = [paramTree valueForKey: @"cutoff"];
        resonanceParameter = [paramTree valueForKey: @"resonance"];
        
        // prevent retain cycle in parameter observer
        __weak FilterDemoViewController *weakSelf = self;
        __weak AUParameter *weakCutoffParameter = cutoffParameter;
        __weak AUParameter *weakResonanceParameter = resonanceParameter;
        parameterObserverToken = [paramTree tokenByAddingParameterObserver:^(AUParameterAddress address, AUValue value) {
            __strong FilterDemoViewController *strongSelf = weakSelf;
            __strong AUParameter *strongCutoffParameter = weakCutoffParameter;
            __strong AUParameter *strongResonanceParameter = weakResonanceParameter;
            
            dispatch_async(dispatch_get_main_queue(), ^{
                if (address == strongCutoffParameter.address){
                    strongSelf->filterView.frequency = value;
                    [strongSelf updateLabelsWithFrequency:[strongCutoffParameter stringFromValue: nil] resonance:nil];

                } else if (address == strongResonanceParameter.address) {
                    strongSelf->filterView.resonance = value;
                    [strongSelf updateLabelsWithFrequency:nil resonance:[strongResonanceParameter stringFromValue: nil]];
                }
                
                [strongSelf updateFilterViewFrequencyAndMagnitudes];
            });
        }];
        
        filterView.frequency = cutoffParameter.value;
        filterView.resonance = resonanceParameter.value;

        [self updateLabelsWithFrequency:[cutoffParameter stringFromValue: nil] resonance:[resonanceParameter stringFromValue: nil]];

        [_audioUnit addObserver:self forKeyPath:@"allParameterValues"
                            options:NSKeyValueObservingOptionNew
                            context:parameterObserverToken];

        // Initial drawing of the graph
        [self updateFilterViewFrequencyAndMagnitudes];
    } else {
        NSLog(@"paramTree is NULL!\n");
    }
}

- (void)disconnectViewWithAU {
    if (parameterObserverToken) {
        [_audioUnit.parameterTree removeParameterObserver:parameterObserverToken];
        [_audioUnit removeObserver:self forKeyPath:@"allParameterValues" context:parameterObserverToken];
        parameterObserverToken = 0;
    }
}

#pragma mark -

- (void)updateFilterViewFrequencyAndMagnitudes {
    if (!_audioUnit) return;
    
    NSArray *frequencies = [filterView frequencyDataForDrawing];
    NSArray *magnitudes  = [_audioUnit magnitudesForFrequencies:frequencies];
    
    [filterView setMagnitudes: magnitudes];
}

#pragma mark -
#pragma mark: <FilterViewDelegate>

- (void)filterViewDidChange:(FilterView *)sender frequency:(double)frequency {
    cutoffParameter.value = frequency;

    [self updateFilterViewFrequencyAndMagnitudes];
}

- (void)filterViewDidChange:(FilterView *)sender resonance:(double)resonance {
    resonanceParameter.value = resonance;

    [self updateFilterViewFrequencyAndMagnitudes];
}

- (void)filterViewDataDidChange:(FilterView *)sender {
    [self updateFilterViewFrequencyAndMagnitudes];
}

#pragma mark: Actions

- (IBAction)setCutoff:(id)sender {
    if (sender == frequencyLabel)
        cutoffParameter.value = frequencyLabel.floatValue;

    if (sender == frequencyLabelSmallView)
        cutoffParameter.value = frequencyLabelSmallView.floatValue;

    [self updateFilterViewFrequencyAndMagnitudes];
}

- (IBAction)setResonance:(id)sender {
    if (sender == resonanceLabel)
        resonanceParameter.value = resonanceLabel.floatValue;

    if (sender == resonanceLabelSmallView)
        resonanceParameter.value = resonanceLabelSmallView.floatValue;

    [self updateFilterViewFrequencyAndMagnitudes];
}

#pragma mark: React to host calls via AUAudioUnit

- (void)handleSelectViewConfiguration:(AUAudioUnitViewConfiguration *)viewConfiguration
{
    if (viewConfiguration.width == 400 && viewConfiguration.height == 100) {
        // small view requested
        if (smallViewActive)
        {
            return; // already active, nothing to do
        }
    } else {
       // large view requested
       if (!smallViewActive)
       {
           return; // already active, nothing to do
       }
    }

    NSView *viewToAdd = smallViewActive ? largeView : smallView;
    NSView *currentView = smallViewActive ? smallView : largeView;

    dispatch_async(dispatch_get_main_queue(), ^{
        self.view.hidden = YES;

        [currentView removeFromSuperview];
        [self.view addSubview:viewToAdd];

        NSDictionary *views = NSDictionaryOfVariableBindings(viewToAdd);

        if (viewToAdd == largeView) {
            // Add constraints to fill out container completely with no padding
            [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-0-[viewToAdd(>=500)]-0-|" options:0 metrics:nil views:views]];
            [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-0-[viewToAdd(>=800)]-0-|" options:0 metrics:nil views:views]];
        } else {
            // Tha smallview should stay on top, and the margin should expand
            [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-0-[viewToAdd(>=100)]-|" options:0 metrics:nil views:views]];
            [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-0-[viewToAdd(>=400)]-|" options:0 metrics:nil views:views]];
        }

        self.view.hidden = NO;
    });

    smallViewActive = !smallViewActive;
}

@end

File: AudioUnit/Filter/iOS/FilterDemoFramework/FilterDemoViewController.swift
/*
See LICENSE.txt for this sample’s licensing information.

Abstract:
View controller for the FilterDemo audio unit. Manages the interactions between a FilterView and
            the audio unit's parameters.
*/

import UIKit
import CoreAudioKit

public class FilterDemoViewController: AUViewController, FilterViewDelegate, UITextFieldDelegate {
    // MARK: Properties

    @IBOutlet weak var filterView: FilterView!
	@IBOutlet weak var frequencyLabel: UILabel!
	@IBOutlet weak var resonanceLabel: UILabel!

    @IBOutlet weak var frequencyTextField: UITextField!
    @IBOutlet weak var resonanceTextField: UITextField!

    @IBOutlet var largeView: UIView!
    @IBOutlet var smallView: UIView!

    var smallViewActive: Bool = false

    /*
		When this view controller is instantiated within the FilterDemoApp, its
        audio unit is created independently, and passed to the view controller here.
	*/
    public var audioUnit: AUv3FilterDemo? {
        didSet {
			/*
				We may be on a dispatch worker queue processing an XPC request at
                this time, and quite possibly the main queue is busy creating the
                view. To be thread-safe, dispatch onto the main queue.
				
				It's also possible that we are already on the main queue, so to
                protect against deadlock in that case, dispatch asynchronously.
			*/
			DispatchQueue.main.async {
				if self.isViewLoaded {
					self.connectViewWithAU()
				}
			}
        }
    }

    @objc
    public func handleSelectViewConfiguration(_ viewConfiguration: AUAudioUnitViewConfiguration) {
        // If requested configuration is already active, do nothing
        if (viewConfiguration.width > 0) && (viewConfiguration.width < 800 || viewConfiguration.height < 500) {
            if smallViewActive { return }
        } else {
            if !smallViewActive { return }
        }

        self.view.translatesAutoresizingMaskIntoConstraints = true

        let targetView = smallViewActive ? largeView : smallView
        let sourceView = smallViewActive ? smallView : largeView
        let containerView = self.view!

        DispatchQueue.main.async {
            UIView.transition(from: sourceView!, to: targetView!, duration: 0.5,
                              options: [UIViewAnimationOptions.transitionCrossDissolve,
                                        UIViewAnimationOptions.layoutSubviews], completion:nil)

            // If the transition is still in progress, add ourselves to the
            // container view and make sure the layout is right, before the transition finishes
            if targetView?.superview !== self.view {
                self.view.addSubview(targetView!)
            }

            // When the view is removed from the containerView, the constraints are lost,
            // re-add them so we fill out the container
            let views = ["targetView": targetView!]
            containerView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:[targetView]-(0)-|",
                                                                        options: [],
                                                                        metrics: nil, views: views))
            containerView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-(0)-[targetView]",
                                                                        options: [],
                                                                        metrics: nil, views: views))
            containerView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-(0)-[targetView]",
                                                                        options: [],
                                                                        metrics: nil, views: views))
            containerView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:[targetView]-(0)-|",
                                                                        options: [],
                                                                        metrics: nil, views: views))
        }
        smallViewActive = !smallViewActive
    }

    var cutoffParameter: AUParameter?
	var resonanceParameter: AUParameter?
	var parameterObserverToken: AUParameterObserverToken?

	public override func viewDidLoad() {
		super.viewDidLoad()

		// Respond to changes in the filterView (frequency and/or response changes).
        filterView.delegate = self

        self.frequencyTextField.delegate = self
        self.resonanceTextField.delegate = self

        guard audioUnit != nil else { return }

        self.view.layer.borderColor = UIColor.black.cgColor
        self.view.layer.borderWidth = 2
        self.view.layer.cornerRadius = 1
        self.view.backgroundColor = UIColor.green

        connectViewWithAU()
	}

    public func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        self.view.endEditing(true)
        return false
    }

    // MARK: FilterViewDelegate

    func updateFilterViewFrequencyAndMagnitudes() {
        guard let audioUnit = audioUnit as AUv3FilterDemo! else { return }

        // Get an array of frequencies from the view.
        let frequencies = filterView.frequencyDataForDrawing()

        // Get the corresponding magnitudes from the AU.
        let magnitudes = audioUnit.magnitudes(forFrequencies: frequencies as [NSNumber]!).map { $0.doubleValue }

        filterView.setMagnitudes(magnitudes)
    }

    func filterView(_ filterView: FilterView, didChangeResonance resonance: Float) {

        resonanceParameter?.value = resonance

        updateFilterViewFrequencyAndMagnitudes()
    }

    func filterView(_ filterView: FilterView, didChangeFrequency frequency: Float) {

        cutoffParameter?.value = frequency

        updateFilterViewFrequencyAndMagnitudes()
    }

    func filterView(_ filterView: FilterView, didChangeFrequency frequency: Float, andResonance resonance: Float) {

        cutoffParameter?.value = frequency
        resonanceParameter?.value = resonance

        updateFilterViewFrequencyAndMagnitudes()
    }

    func filterViewDataDidChange(_ filterView: FilterView) {
        updateFilterViewFrequencyAndMagnitudes()
    }

    func updateLabels(resonance: String?, cutoff: String?) {
        if resonance != nil {
            resonanceLabel.text = resonance
            resonanceTextField.text = resonance
        }
        if cutoff != nil {
            frequencyLabel.text = cutoff
            frequencyTextField.text = cutoff
        }
    }

    @IBAction func frequencyUpdated(_ sender : Any) {
        guard let paramTree = audioUnit?.parameterTree else { return }
        cutoffParameter = paramTree.value(forKey: "cutoff") as? AUParameter
        cutoffParameter!.value = (frequencyTextField.text! as NSString).floatValue
        frequencyTextField.text = cutoffParameter!.string(fromValue: nil)
    }

    @IBAction func resonanceUpdated(_ sender: Any) {
        guard let paramTree = audioUnit?.parameterTree else { return }
        resonanceParameter = paramTree.value(forKey: "resonance") as? AUParameter
        resonanceParameter!.value = (resonanceTextField.text! as NSString).floatValue
        resonanceTextField.text = resonanceParameter!.string(fromValue: nil)
    }

    /*
		We can't assume anything about whether the view or the AU is created first.
		This gets called when either is being created and the other has already
        been created.
	*/
	func connectViewWithAU() {
		guard let paramTree = audioUnit?.parameterTree else { return }

        audioUnit?.filterDemoViewController = self

		cutoffParameter = paramTree.value(forKey: "cutoff") as? AUParameter
		resonanceParameter = paramTree.value(forKey: "resonance") as? AUParameter

        parameterObserverToken = paramTree.token(byAddingParameterObserver: { [weak self] address, value in
            guard let strongSelf = self else { return }

			DispatchQueue.main.async {
				if address == strongSelf.cutoffParameter!.address {
					strongSelf.filterView.frequency = value
					strongSelf.updateLabels(resonance: nil, cutoff: strongSelf.cutoffParameter!.string(fromValue: nil))
				} else if address == strongSelf.resonanceParameter!.address {
					strongSelf.filterView.resonance = value
					strongSelf.updateLabels(resonance: strongSelf.resonanceParameter!.string(fromValue: nil), cutoff: nil)
				}

				strongSelf.updateFilterViewFrequencyAndMagnitudes()
			}
		})

        filterView.frequency = cutoffParameter!.value
        filterView.resonance = resonanceParameter!.value

        updateFilterViewFrequencyAndMagnitudes()
        updateLabels(resonance: resonanceParameter!.string(fromValue: nil),
                     cutoff: cutoffParameter!.string(fromValue: nil))
	}
}
